sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerTransformRegistrar.java

```diff
@@ -0,0 +1,287 @@
+          checkArgument(
+              timeUnit != null,
+              "Time unit cannot be null when MAX_STALENESS or EXACT_STALENESS mode is selected");
+        }
+        if (mode == TimestampBound.Mode.READ_TIMESTAMP
+            || mode == TimestampBound.Mode.MIN_READ_TIMESTAMP) {
+          checkArgument(
+              readTimestamp != null,
+              "Timestamp cannot be null when READ_TIMESTAMP or MIN_READ_TIMESTAMP mode is selected");
+        }
+        switch (mode) {
+          case STRONG:
+            return TimestampBound.strong();
+          case MAX_STALENESS:
+            return TimestampBound.ofMaxStaleness(exactStaleness, TimeUnit.valueOf(timeUnit));
+          case EXACT_STALENESS:
+            return TimestampBound.ofExactStaleness(exactStaleness, TimeUnit.valueOf(timeUnit));
+          case READ_TIMESTAMP:
+            return TimestampBound.ofReadTimestamp(Timestamp.parseTimestamp(readTimestamp));
+          case MIN_READ_TIMESTAMP:
+            return TimestampBound.ofMinReadTimestamp(Timestamp.parseTimestamp(readTimestamp));
+          default:
+            throw new RuntimeException("Unknown timestamp bound mode: " + mode);
+        }
+      }
+
+      public ReadOperation getReadOperation() {
+        checkArgument(
+            sql == null || table == null,
+            "Query and table params are mutually exclusive. Set just one of them.");
+        if (sql != null) {
+          return ReadOperation.create().withQuery(sql);
+        }
+        return ReadOperation.create().withTable(table).withColumns(schema.getFieldNames());
+      }
+    }
+
+    @Override
+    public PTransform<PBegin, PCollection<Row>> buildExternal(Configuration configuration) {
+      SpannerIO.Read readTransform =
+          SpannerIO.read()
+              .withProjectId(configuration.projectId)
+              .withDatabaseId(configuration.databaseId)
+              .withInstanceId(configuration.instanceId)
+              .withReadOperation(configuration.getReadOperation());
+
+      if (configuration.host != null) {
+        readTransform = readTransform.withHost(configuration.host);
+      }
+      if (configuration.emulatorHost != null) {
+        readTransform = readTransform.withEmulatorHost(configuration.emulatorHost);
+      }
+      if (configuration.getTimestampBound() != null) {
+        readTransform = readTransform.withTimestampBound(configuration.getTimestampBound());
+      }
+      if (configuration.batching != null) {
+        readTransform = readTransform.withBatching(configuration.batching);
+      }
+
+      return new SpannerIO.ReadRows(readTransform, configuration.schema);
+    }
+  }
+
+  @Experimental(Kind.PORTABILITY)
+  public static class WriteBuilder
+      implements ExternalTransformBuilder<WriteBuilder.Configuration, PCollection<Row>, PDone> {
+
+    public static class Configuration extends CrossLanguageConfiguration {
```

Comment:
FYI now with https://github.com/apache/beam/pull/12481 it's possible to use schema inference for these configuration objects, so you should be able to use `@AutoValue` and `AutoValueSchema` which could save a lot of boiler plate There's an example here:

https://github.com/apache/beam/blob/89a2d17624a5f2f445b7199fe7a61ec0eca8205a/sdks/java/expansion-service/src/test/java/org/apache/beam/sdk/expansion/service/ExpansionServiceTest.java#L305-L315

(it's fine to leave it as-is, just letting you know)

Reply:
Many setters have some logic inside them and AutoValue does not allow inheritance so there would be some code duplication in getter methods. As this is just a cosmetic thing and I have other things to do I'll leave it as it is for now.

Reply:
Sounds good